app/src/main/java/com/concordium/wallet/core/gson/NullableTypeAdapterFactory.kt

```diff
@@ -0,0 +1,52 @@
+package com.concordium.wallet.core.gson
+
+import com.google.gson.*
+import com.google.gson.reflect.TypeToken
+import com.google.gson.stream.JsonReader
+import com.google.gson.stream.JsonWriter
+import kotlin.jvm.internal.Reflection
+import kotlin.reflect.KClass
+import kotlin.reflect.full.memberProperties
+
+/**
+ * GSON doesn't understand Kotlin’s non-null types.
+ *
+ * This adapter adds the ability to throw an error when a non-null value is set to null.
+ * @return [TypeAdapterFactory]
+ *
+ * [JsonParseException] if a non-null value is null
+ */
+class NullableTypeAdapterFactory : TypeAdapterFactory {
+
+    override fun <T : Any> create(gson: Gson, type: TypeToken<T>): TypeAdapter<T>? {
+        val delegate = gson.getDelegateAdapter(this, type)
+
+        // If the class isn't kotlin, don't use the custom type adapter
+        if (type.rawType.declaredAnnotations.none { it.annotationClass.qualifiedName == "kotlin.Metadata" }) {
+            return null
+        }
+
+        return object : TypeAdapter<T>() {
+
+            override fun write(out: JsonWriter, value: T?) = delegate.write(out, value)
+
+            override fun read(input: JsonReader): T? {
+                val value: T? = delegate.read(input)
+
+                if (value != null) {
+                    val kotlinClass: KClass<Any> = Reflection.createKotlinClass(type.rawType)
```

Comment:
This looks potentially expensive. Can you do it just after the kotlin class check above? Then most of the work is only done once rather than on every read. You could extract the reflection objects of the nullable fields and then in `read` you only check if any of these fields are `null`.